internal/cloud/backend.go

```diff
@@ -424,40 +435,44 @@ func (b *Cloud) setConfigurationFields(obj cty.Value) tfdiags.Diagnostics {
 		b.organization = val.AsString()
 	}
 
-	// Initially set the project via env var
-	b.WorkspaceMapping.Project = os.Getenv("TF_CLOUD_PROJECT")
+	// Initially, set workspaces from the configuration
+	b.WorkspaceMapping = newWorkspacesMappingFromFields(obj)
 
-	// Initially set the workspace name via env var
-	b.WorkspaceMapping.Name = os.Getenv("TF_WORKSPACE")
+	// Overwrite workspaces config from env variable
+	if diag := reconcileWorkspaceMappingEnvVars(&b.WorkspaceMapping); diag != nil {
+		return diags.Append(diag)
+	}
 
-	// Get the workspaces configuration block and retrieve the
-	// default workspace name.
-	if workspaces := obj.GetAttr("workspaces"); !workspaces.IsNull() {
+	// Determine if we are forced to use the local backend.
+	b.forceLocal = os.Getenv("TF_FORCE_LOCAL_BACKEND") != ""
 
-		// Check if the project is present and valid in the config.
-		if val := workspaces.GetAttr("project"); !val.IsNull() && val.AsString() != "" {
-			b.WorkspaceMapping.Project = val.AsString()
-		}
+	return diags
+}
 
-		// PrepareConfig checks that you cannot set both of these.
-		if val := workspaces.GetAttr("name"); !val.IsNull() {
-			b.WorkspaceMapping.Name = val.AsString()
+func reconcileWorkspaceMappingEnvVars(w *WorkspaceMapping) tfdiags.Diagnostic {
+	// See: https://github.com/opentofu/opentofu/issues/814
+	if v := os.Getenv("TF_WORKSPACE"); v != "" && w.Name == "" {
+		if len(w.Tags) > 0 && !workspaceInTags(w.Tags, v) {
+			return invalidWorkspaceConfigMisconfigurationEnvVar
```

Comment:
I think we just want to only use the workspace and tag env vars if the workspace block isn't there, no? Basically, only look at env var if workspace configuration isn't in code.

Reply:
It's slightly misleading because it'd contradict to the behaviour illustrated in the documentation:

> The name of a single cloud backend workspace. OpenTofu reads this when workspaces is omitted from the cloud block. The cloud backend will not create a new workspace from this variable; the workspace must exist in the specified organization. You can set TF_WORKSPACE if the cloud block uses tags. However, the value of TF_WORKSPACE must be included in the set of tags.

It's a bit cryptic, however, I read it as following: `TF_WORKSPACE` is used in two cases:
- `workspaces` config is not present in the `cloud` block;
- `workspaces.tags` is set only.

I find the logic implemented in 1.5.7 which we agreed to use a reference contradicting that understanding, hence I adjusted it accordingly.

WDYT?

Reply:
You're explanation, and especially the docs excerpt clear things up. LGTM, thanks!